Epmlstrcmw 28 Float to BigDecimal

SergeySdv opened the pull request:
Now WomFloat set BigDecimal as type for big numbers.

Commit:
Add Changed type Double to BigDecimal

svitkovsergey commented on wdl/transforms/new-base/src/main/scala/wdl/transforms/base/linking/expression/values/EngineFunctionEvaluators.scala:
```diff
@@ -460,8 +462,10 @@ object EngineFunctionEvaluators {
+      import BigDecimal.RoundingMode.CEILING
+      def ceil(a: BigDecimal): BigDecimal = a.setScale(0, CEILING)
+      processValidatedSingleValue[WomFloat, WomInteger](a.param.evaluateValue(inputs, ioFunctionSet, forCommandInstantiationOptions)) { bigDecimal =>
+        EvaluatedValue(WomInteger(ceil(bigDecimal.value).toInt), Seq.empty).validNel
```
What will happen if `bigDecimal` would contain a value outside `Int` range?

svitkovsergey commented on centaur/src/main/resources/reference.conf:
```diff
@@ -47,7 +47,7 @@ centaur {
   # case will be in a subdirectory named FOO with workflow, inputs, and options files.
   standardTestCasePath: "centaur/src/main/resources/standardTestCases"
-  # optionalTestPath: "/some/path/to/tests"
+  optionalTestPath: "centaur/src/main/resources/integrationTestCases"
```
Are you sure that this change is needed?

DenisVerkhoturov replied:
@SergeySdv is it not going to work if you do not specify this option?

svitkovsergey commented on supportedBackends/google/pipelines/v2alpha1/src/main/scala/cromwell/backend/google/pipelines/v2alpha1/MachineConstraints.scala:
```diff
@@ -8,10 +8,18 @@ import mouse.all._
+      else throw TypeCastMismatch("Can not allocate memory")
+    }
+
+    def asMultipleOf(factor: MemorySize): MemorySize = MemorySize(factor.amount * ceil(information.bytes / factor.bytes), factor.unit)
```
This line is 130 character long, consider splitting it into two lines.

svitkovsergey commented on wdl/model/draft2/src/main/scala/wdl/draft2/model/expression/WdlStandardLibraryFunctions.scala:
```diff
@@ -129,15 +130,19 @@ trait WdlStandardLibraryFunctions extends WdlFunctions[WomValue] {
   def floor(params: Seq[Try[WomValue]]): Try[WomInteger] = {
-    extractSingleArgument("floor", params) flatMap { f => WomFloatType.coerceRawValue(f) } map { f => WomInteger(Math.floor(f.asInstanceOf[WomFloat].value).toInt) }
+    def floor(a: BigDecimal): BigDecimal = a.setScale(0, FLOOR)
+    extractSingleArgument("floor", params) flatMap { f => WomFloatType.coerceRawValue(f) } map { f => WomInteger(floor(f.asInstanceOf[WomFloat].value).toInt) }
```
This line is 155 character long, consider splitting it into two lines.

svitkovsergey commented on wdl/model/draft2/src/main/scala/wdl/draft2/model/expression/WdlStandardLibraryFunctions.scala:
```diff
@@ -129,15 +130,19 @@ trait WdlStandardLibraryFunctions extends WdlFunctions[WomValue] {
   def round(params: Seq[Try[WomValue]]): Try[WomInteger] = {
-    extractSingleArgument("round", params) flatMap { f => WomFloatType.coerceRawValue(f) } map { f => WomInteger(Math.round(f.asInstanceOf[WomFloat].value).toInt) }
+    def round(a: BigDecimal): BigDecimal = a.setScale(0, HALF_UP)
+    extractSingleArgument("round", params) flatMap { f => WomFloatType.coerceRawValue(f) } map { f => WomInteger(round(f.asInstanceOf[WomFloat].value).toInt) }
```
This line is 155 character long, consider splitting it into two lines.

svitkovsergey commented on wdl/model/draft2/src/main/scala/wdl/draft2/model/expression/WdlStandardLibraryFunctions.scala:
```diff
@@ -129,15 +130,19 @@ trait WdlStandardLibraryFunctions extends WdlFunctions[WomValue] {
-    extractSingleArgument("ceil", params) flatMap { f => WomFloatType.coerceRawValue(f) } map { f => WomInteger(Math.ceil(f.asInstanceOf[WomFloat].value).toInt) }
+
+    def ceil(a: BigDecimal): BigDecimal = a.setScale(0, CEILING)
+    extractSingleArgument("ceil", params) flatMap { f => WomFloatType.coerceRawValue(f) } map { f => WomInteger(ceil(f.asInstanceOf[WomFloat].value).toInt) }
```
This line is 153 character long, consider splitting it into two lines.

svitkovsergey commented on wdl/transforms/new-base/src/main/scala/wdl/transforms/base/linking/expression/values/EngineFunctionEvaluators.scala:
```diff
@@ -460,8 +462,10 @@ object EngineFunctionEvaluators {
-        EvaluatedValue(WomInteger(math.ceil(float.value).toInt), Seq.empty).validNel
+      import BigDecimal.RoundingMode.CEILING
+      def ceil(a: BigDecimal): BigDecimal = a.setScale(0, CEILING)
+      processValidatedSingleValue[WomFloat, WomInteger](a.param.evaluateValue(inputs, ioFunctionSet, forCommandInstantiationOptions)) { bigDecimal =>
```
This line is 143 character long, consider splitting it into two lines.

SergeySdv replied:
Fixed

svitkovsergey commented on wdl/transforms/new-base/src/main/scala/wdl/transforms/base/linking/expression/values/EngineFunctionEvaluators.scala:
```diff
@@ -448,8 +448,10 @@ object EngineFunctionEvaluators {
-        EvaluatedValue(WomInteger(math.floor(float.value).toInt), Seq.empty).validNel
+      import BigDecimal.RoundingMode.FLOOR
+      def floor(a: BigDecimal): BigDecimal = a.setScale(0, FLOOR)
+      processValidatedSingleValue[WomFloat, WomInteger](a.param.evaluateValue(inputs, ioFunctionSet, forCommandInstantiationOptions)) { bidDecimal =>
```
This line is 143 character long, consider splitting it into two lines.

SergeySdv replied:
Fixed

svitkovsergey commented on wdl/transforms/new-base/src/main/scala/wdl/transforms/base/linking/expression/values/EngineFunctionEvaluators.scala:
```diff
@@ -472,8 +476,10 @@ object EngineFunctionEvaluators {
-        EvaluatedValue(WomInteger(math.round(float.value).toInt), Seq.empty).validNel
+      import BigDecimal.RoundingMode.HALF_UP
+      def round(a: BigDecimal): BigDecimal = a.setScale(0, HALF_UP)
+      processValidatedSingleValue[WomFloat, WomInteger](a.param.evaluateValue(inputs, ioFunctionSet, forCommandInstantiationOptions)) { bigDecimal =>
```
This line is 143 character long, consider splitting it into two lines.

SergeySdv replied:
Done

svitkovsergey commented on centaur/src/main/resources/integrationTestCases/largenumber.test:
```diff
@@ -0,0 +1,12 @@
+metadata {
+  workflowName: Largenumber
+  status: Succeeded
+}
```
Please check "ensure line feed at file end on save" in IDE settings.

svitkovsergey commented on cwl/src/main/scala/cwl/internal/CwlEcmaScriptDecoder.scala:
```diff
@@ -38,7 +38,8 @@ class CwlEcmaScriptDecoder {
         WomInteger(double.intValue).valid
-      case double: java.lang.Double => WomFloat(double).valid
+      case double: java.lang.Double => WomFloat(BigDecimal(double)).valid
+      case bigDecimal: java.math.BigDecimal => WomFloat(bigDecimal).valid
```
Why `java.math.BigDecimal`, not `scala.math.BigDecimal`? 
I also think that it would be better to import this class and remove path declaration. Same applies to other classes here.

SergeySdv replied:
Actually should i change code style in this class in my request if it does not matter for my current task?

svitkovsergey replied:
@SergeySdv https://www.matheus.ro/2017/12/11/clean-code-boy-scout-rule/

DenisVerkhoturov replied:
@likeanowl, As far as I can see such style was there before, no reason to change it in this PR for the only two `case`-s in the entire `match`, I guess.

svitkovsergey commented on supportedBackends/google/pipelines/v2alpha1/src/main/scala/cromwell/backend/google/pipelines/v2alpha1/MachineConstraints.scala:
```diff
@@ -8,10 +8,18 @@ import mouse.all._
 import wom.format.MemorySize
+import scala.math.BigDecimal.RoundingMode.CEILING
+
+case class TypeCastMismatch(message: String) extends Exception(message)
```
I think that this exception class should have more informative name, something like `NotInDoubleRangeException`.

svitkovsergey commented on supportedBackends/sfs/src/main/scala/cromwell/backend/impl/sfs/config/MemoryDeclarationValidation.scala:
```diff
@@ -53,11 +53,12 @@ class MemoryDeclarationValidation(declaration: Declaration, attributeName: Strin
     validation.withDefault(WomLong(memorySize.bytes.toLong))
   }
 
+  case class TypeCastMismatch(message: String) extends Exception(message)
```
Shouldn't we delare this exception class in some common module in order to keep DRY?

DenisVerkhoturov replied:
I agree with @likeanowl and [his another comment](https://github.com/EpamLifeSciencesTeam/cromwell/pull/19/files#r325709371) about the same class, since the only reason this exception is used for is the validation logic of whether the value is a decimal double it would be more informative to name it accordingly such as @likeanowl suggested or even more straightforward, for instance, `NotADecimalDoubleException`.

Concerning the same class defined in the `MachineConstraints` i suggest removing it and the entire method `ceil`.

svitkovsergey commented on supportedBackends/sfs/src/main/scala/cromwell/backend/impl/sfs/config/MemoryDeclarationValidation.scala:
```diff
@@ -53,11 +53,12 @@ class MemoryDeclarationValidation(declaration: Declaration, attributeName: Strin
       case WomInteger(value) => value.toDouble
       case WomLong(value) => value.toDouble
-      case WomFloat(value) => value
+      case WomFloat(value) => if(value.isDecimalDouble) value.doubleValue else throw TypeCastMismatch("Can not allocate memory")
```
Extra space is required after `if`.

svitkovsergey reviewed (changes requested):
Great request! 👍 
I left a couple of comments below. Also, shouldn't we also add an extra matching case to `WomFloatType.coerceRawValue` function?

iliaonishchenko reviewed (commented):
Agree with @likeanowl

myazinn commented on wdl/model/draft2/src/main/scala/wdl/draft2/model/expression/WdlStandardLibraryFunctions.scala:
```diff
@@ -129,15 +130,19 @@ trait WdlStandardLibraryFunctions extends WdlFunctions[WomValue] {
 
   def floor(params: Seq[Try[WomValue]]): Try[WomInteger] = {
-    extractSingleArgument("floor", params) flatMap { f => WomFloatType.coerceRawValue(f) } map { f => WomInteger(Math.floor(f.asInstanceOf[WomFloat].value).toInt) }
+    def floor(a: BigDecimal): BigDecimal = a.setScale(0, FLOOR)
```
Although it works fine, I find confusing to have two methods with the same name but different outputs (and inputs).
I think you should use another name that will express the difference between methods.

svitkovsergey replied:
Good point! I didn't notice that during my brief review of this PR.
@SergeySdv Is there a need for this inner method at all? I think that `bd.setScale(0, FLOOR)` is already pretty short and expressive.

myazinn commented on wdl/model/draft2/src/main/scala/wdl/draft2/model/expression/WdlStandardLibraryFunctions.scala:
```diff
@@ -129,15 +130,19 @@ trait WdlStandardLibraryFunctions extends WdlFunctions[WomValue] {
   def ceil(params: Seq[Try[WomValue]]): Try[WomInteger] = {
-    extractSingleArgument("ceil", params) flatMap { f => WomFloatType.coerceRawValue(f) } map { f => WomInteger(Math.ceil(f.asInstanceOf[WomFloat].value).toInt) }
+
+    def ceil(a: BigDecimal): BigDecimal = a.setScale(0, CEILING)
```
Same problem as on line 133

myazinn commented on wdl/model/draft2/src/main/scala/wdl/draft2/model/expression/WdlStandardLibraryFunctions.scala:
```diff
@@ -129,15 +130,19 @@ trait WdlStandardLibraryFunctions extends WdlFunctions[WomValue] {
 
   def round(params: Seq[Try[WomValue]]): Try[WomInteger] = {
-    extractSingleArgument("round", params) flatMap { f => WomFloatType.coerceRawValue(f) } map { f => WomInteger(Math.round(f.asInstanceOf[WomFloat].value).toInt) }
+    def round(a: BigDecimal): BigDecimal = a.setScale(0, HALF_UP)
```
Same problem as on line 133.

myazinn commented on supportedBackends/google/pipelines/v2alpha1/src/main/scala/cromwell/backend/google/pipelines/v2alpha1/MachineConstraints.scala:
```diff
@@ -8,10 +8,18 @@ import mouse.all._
 object MachineConstraints {
   implicit class EnhancedInformation(val information: MemorySize) extends AnyVal {
-    def asMultipleOf(factor: MemorySize): MemorySize = MemorySize(factor.amount * (information.bytes / factor.bytes).ceil, factor.unit)
+    def ceil(a: BigDecimal): Double = {
```
It's quite strange that a method called `ceil` returns the value converted from BigDecimal to Double.
Consider a more appropriate name.

myazinn commented on supportedBackends/google/pipelines/v2alpha1/src/main/scala/cromwell/backend/google/pipelines/v2alpha1/MachineConstraints.scala:
```diff
@@ -8,10 +8,18 @@ import mouse.all._
+      else throw TypeCastMismatch("Can not allocate memory")
+    }
+
+    def asMultipleOf(factor: MemorySize): MemorySize = MemorySize(factor.amount * ceil(information.bytes / factor.bytes), factor.unit)
```
Both `information.bytes` and `factor.bytes` are Double. Why are you ceiling it with a function that takes BigDecimal as an argument?

DenisVerkhoturov replied:
The previuos version of this logic was:
```scala
MemorySize(factor.amount * (information.bytes / factor.bytes).ceil, factor.unit)
```
@SergeySdv, why do we need to define and use our own method `ceil: [BigDecimal => Double]` even though `MemorySize.bytes` is of type `Double`, as @myazinn mentioned?

svitkovsergey commented:
Also, @SergeySdv I just noticed that your branch name contains a typo: `EPMLSTRCMW-28_float_to_bigDecamal`. Please consider fixing it.

svitkovsergey commented:
@SergeySdv I also noticed that your commit message does not follows our commit messages guidelines. Please consider reading this https://udacity.github.io/git-styleguide/ and this https://chris.beams.io/posts/git-commit/ guides and change your commit message according to them.

DenisVerkhoturov commented on wom/src/main/scala/wom/values/WomString.scala:
```diff
@@ -38,4 +38,4 @@ case class WomString(value: String) extends WomPrimitive {
   override def toWomString = "\"" + StringEscapeUtils.escapeJava(value) + "\""
   override def valueString = value
-}
+}
```
Please, add a line break at the end of thit file.

DenisVerkhoturov reviewed (changes requested):
Thank you for the pull request! 👍

Commit:
feat: Change WomFloat Double type implementation to BigDecimal

This small change will help to store larger values in WomFloat type.
Also added validation logic of type compatibility.

myazinn reviewed (approved):
Much better

svitkovsergey commented on wdl/transforms/new-base/src/main/scala/wdl/transforms/base/linking/expression/values/EngineFunctionEvaluators.scala:
```diff
@@ -243,7 +244,7 @@ object EngineFunctionEvaluators {
 
         tryResult.map(v => EvaluatedValue(v, Seq(CommandSetupSideEffectFile(v)))).toErrorOr.contextualizeErrors(s"""$functionName(...)""")
-      } (coercer = WomArrayType(WomStringType))
+      }(coercer = WomArrayType(WomStringType))
```
Should this space really be removed?

SergeySdv replied:
I actually add it previously when  i was reading code for fixing. And this space was not exists before, so i cleared all unnecessary changes as Denis mentioned.

svitkovsergey replied:
As I can see, this space exists in cromwell's develop branch: https://github.com/broadinstitute/cromwell/blob/70cff69799f149191bdaec5d8878dd2a3d5202b7/wdl/transforms/new-base/src/main/scala/wdl/transforms/base/linking/expression/values/EngineFunctionEvaluators.scala#L246

svitkovsergey commented on wdl/transforms/new-base/src/main/scala/wdl/transforms/base/linking/expression/values/EngineFunctionEvaluators.scala:
```diff
@@ -262,7 +263,7 @@ object EngineFunctionEvaluators {
 
         tryResult.map(v => EvaluatedValue(v, Seq(CommandSetupSideEffectFile(v)))).toErrorOr.contextualizeErrors(s"""$functionName(...)""")
-      } (coercer = WomArrayType(WomAnyType))
+      }(coercer = WomArrayType(WomAnyType))
```
same as at line 247

svitkovsergey replied:
https://github.com/broadinstitute/cromwell/blob/70cff69799f149191bdaec5d8878dd2a3d5202b7/wdl/transforms/new-base/src/main/scala/wdl/transforms/base/linking/expression/values/EngineFunctionEvaluators.scala#L265

svitkovsergey commented:
@SergeySdv please take a look at https://github.com/broadinstitute/cromwell/blob/6e76a0e8d8e7a1c70ed18c2dff30307832866453/wom/src/main/scala/wom/types/WomFloatType.scala#L11

SergeiIonin reviewed (approved):
I'll ask some questions later in Slack, I don't think they will affect the approval anyway

Commit:
fix:
 * Add type match for WomFloat to BigDecimal
 * Long lines splitting
